src/trainer.py

```diff
@@ -1115,14 +1137,67 @@ def _save_checkpoint(self, training_state, phase="pretrain", new_best_macro=Fals
                 "metric_state_{}_epoch_{}{}.th".format(phase, epoch, best_str),
             ),
         )
-        log.info("Saved to %s", self._serialization_dir)
+        log.info("Saved general checkpoints to %s", self._serialization_dir)
+
+        # For each of the tasks, we update the best task-speciic checkpoints
+        # only for target training.
+        if phase == "target_train":
+            self.save_task_specific_checkpoints(epoch, phase, task_states, model_state)
 
         if new_best_macro:
             self._unmark_previous_best(phase, epoch)
 
         if not self._keep_all_checkpoints:
             self._delete_old_checkpoints(phase, epoch)
 
+    def save_task_specific_checkpoints(self, epoch, phase, task_states, model_state):
+        """
+        Saves task specific checkpoints for all tasks where the validation performance improved. 
+
+        Parameters 
+        --------------------
+            - epoch: int
+            - task_states: dict, dictionary with task experiment parameters such as optimizer and scheduler. 
+            - model_state: experiment model
+        
+        Returns
+        --------------------
+        None
+        """
+        for metric_name, metric_info in self._metric_infos.items():
```

Comment:
I don't follow the logic here—it seems like we're marking everything as best.

Reply:
We only mark a task's metric_infos as best only if the validation metrics during that epoch is better than the previous validation passes.

Reply:
All of the strings below specify 'best'. Where are the regular non-bests saved?

Reply:
They're stored in the same directory file.

Reply:
I'm asking for where in the code. The name of this method doesn't match what it does.

In any case, if this is meant for pretraining, we may not need it at all...

Reply:
It's here: https://github.com/nyu-mll/jiant/blob/15dd439fd9b998f136a6d8e865d43a6ec21a2c05/src/trainer.py#L1035
We unmark the previous "best" checkpoint before we save a best checkpoint

Reply:
But what if we don't hit a new best at all, it has just been 1000 steps (or similar) and we want to checkpoint anyway? In the current code, we will save such checkpoints. The title of this method makes me think that should live here, but it doesn't seem to.

Reply:
I think I see now—this looks reasonable as-is.